File: device/cuda/hawkes/native/kernel.c
#include "kernel.h"
#include "hawkes_dispatch.h"
#include "../internal/bridge/core_private.h"

int cuda_dispatch_hawkes_kernel_matrix(
    CUDADeviceRef contextRef,
    int elementDType,
    CUDABufferRef eventsRef,
    CUDABufferRef alphaRef,
    CUDABufferRef betaRef,
    CUDABufferRef outRef,
    uint32_t eventCount,
    uint64_t completionToken,
    CUDAStatus* status
) {
    cuda_status_clear(status);

    if (eventCount == 0) {
        return 0;
    }

    if (eventsRef == NULL || alphaRef == NULL || betaRef == NULL || outRef == NULL) {
        cuda_status_set(status, -2, "nil CUDA Hawkes kernel matrix buffer");
        return -2;
    }

    char kernelName[128];
    int nameCode = cuda_hm_kernel_name(kernelName, sizeof(kernelName), "hawkes_kernel_matrix", elementDType, status);

    if (nameCode != 0) {
        return nameCode;
    }

    CUDAContext* context = NULL;
    CUDAStreamRef stream = NULL;
    CUDAKernelRef kernel = NULL;
    int prepareCode = cuda_hm_get_kernel(contextRef, kernelName, status, &context, &stream, &kernel);

    if (prepareCode != 0) {
        return prepareCode;
    }

    void* eventsPtr = cuda_buffer_device_ptr(eventsRef);
    void* alphaPtr = cuda_buffer_device_ptr(alphaRef);
    void* betaPtr = cuda_buffer_device_ptr(betaRef);
    void* outPtr = cuda_buffer_device_ptr(outRef);
    void* args[] = {&eventsPtr, &alphaPtr, &betaPtr, &outPtr, &eventCount};
    uint32_t total = eventCount * eventCount;
    int launchCode = cuda_launch_1d(context, kernel, stream, total, args, sizeof(args), status);

    if (launchCode != 0) {
        return launchCode;
    }

    cuda_track_completion(context, stream, completionToken, NULL, status);
    return 0;
}

File: device/metal/hawkes/native/kernel.m
#include "kernel.h"
#include "hawkes.h"
#include "../internal/bridge/core_private.h"

#include <Foundation/Foundation.h>
#include <Metal/Metal.h>

int metal_dispatch_hawkes_kernel_matrix(
    MetalDeviceRef contextRef,
    int elementDType,
    MetalBufferRef eventsRef,
    MetalBufferRef alphaRef,
    MetalBufferRef betaRef,
    MetalBufferRef outRef,
    uint32_t eventCount,
    uint64_t completionToken,
    MetalStatus* status
) {
    @autoreleasepool {
        metal_hm_status_clear(status);

        if (eventsRef == NULL || alphaRef == NULL || betaRef == NULL || outRef == NULL) {
            metal_hm_status_set(status, -2, "nil Metal buffer");
            return -2;
        }

        id<MTLCommandBuffer> commandBuffer = nil;
        id<MTLComputePipelineState> pipeline = nil;
        int prepareCode = metal_hm_simple_prepare(
            contextRef, "hawkes_kernel_matrix", elementDType, status, &commandBuffer, &pipeline
        );

        if (prepareCode != 0) {
            return prepareCode;
        }

        id<MTLComputeCommandEncoder> encoder = nil;
        int encoderCode = metal_hm_encoder(commandBuffer, pipeline, status, &encoder);
        if (encoderCode != 0) {
            return encoderCode;
        }

        uint32_t total = eventCount * eventCount;
        [encoder setBuffer:(__bridge id<MTLBuffer>)eventsRef offset:0 atIndex:0];
        [encoder setBuffer:(__bridge id<MTLBuffer>)alphaRef offset:0 atIndex:1];
        [encoder setBuffer:(__bridge id<MTLBuffer>)betaRef offset:0 atIndex:2];
        [encoder setBuffer:(__bridge id<MTLBuffer>)outRef offset:0 atIndex:3];
        [encoder setBytes:&eventCount length:sizeof(eventCount) atIndex:4];
        [encoder
            dispatchThreads:MTLSizeMake((NSUInteger)total, 1, 1)
            threadsPerThreadgroup:MTLSizeMake(metal_hm_thread_width(pipeline), 1, 1)
        ];
        metal_track_command_completion((MetalContext*)contextRef, commandBuffer, completionToken, NULL);
        metal_end_encoder((MetalContext*)contextRef, encoder, commandBuffer);

        return 0;
    }
}
